File: go/leet_code/2487.go
// leet_code/2487.go
// This file contains the implementation of a linked list and functions to manipulate it.
// Author: jerodg <https://github.com/jerodg/>

package leet_code

// ListNode represents a node in a singly linked list.
type ListNode struct {
	Val  int       // Val is the value of the node.
	Next *ListNode // Next is a pointer to the next node in the list.
}

// removeNodes removes nodes from the list that have a greater value than the next node.
// It first reverses the list, removes the nodes, and then reverses the list again.
// It returns the head of the modified list.
func removeNodes(head *ListNode) *ListNode {
	head = reverseList(head)
	temp := head
	for temp != nil {
		if temp.Next != nil && temp.Val > temp.Next.Val {
			temp.Next = temp.Next.Next
		} else {
			temp = temp.Next
		}
	}
	return reverseList(head)
}

// reverseList reverses the order of nodes in the list.
// It returns the head of the reversed list.
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		next := head.Next
		head.Next = prev
		prev, head = head, next
	}
	return prev
}

File: cpp/leet_code/2487.cpp
#pragma GCC optimize("O3", "unroll-loops")

#include <ios> // Provides the core input/output library

// ListNode is a struct that represents a node in a singly linked list.
// Each node contains an integer value and a pointer to the next node in the
// list.
struct ListNode {
  int val;        // The integer value stored in this node
  ListNode *next; // Pointer to the next node in the list

  // Default constructor initializes the node with a value of 0 and next pointer
  // as nullptr
  ListNode() : val(0), next(nullptr) {}

  // Constructor initializes the node with a given value and next pointer as
  // nullptr
  explicit ListNode(int x) : val(x), next(nullptr) {}

  // Constructor initializes the node with a given value and a given next node
  ListNode(const int x, ListNode *next) : val(x), next(next) {}
};

// Solution class encapsulates the method to solve the problem
class Solution {
public:
  // Method to remove nodes from a linked list based on a specific condition
  // Parameters:
  // - head: a pointer to the head node of the linked list
  // Returns: a pointer to the head node of the modified linked list
  static ListNode *removeNodes(ListNode *head) {
    std::ios_base::sync_with_stdio(
        false); // Speed up I/O operations by untieing cin from cout

    // If the list is empty, return the empty list
    if (head == nullptr)
      return head;

    // Recursively remove nodes from the rest of the list
    head->next = removeNodes(head->next);

    // If the list has only one node, return the single-node list
    if (head->next == nullptr)
      return head;

    // If the next node's value is greater than the current node's value, remove
    // the current node
    if (head->next->val > head->val)
      return head->next;

    // Otherwise, keep the current node
    return head;
  }
};
